Add API readiness checklist to Release Issue and Review PR (#90)
camaraproject/tooling: release_automation/docs/api-readiness-checklist.md

# API Readiness Checklist

> This document defines the release assets that API teams must provide for each release, and how readiness is verified during the release process.

Reviewer: ```suggestion
This document defines the release assets that API teams must provide for each release depening on the API(s) status(es), and how readiness is verified during the release process.
```

## Purpose

Before an API repository can be released, codeowners must ensure that certain assets are in place and meet quality expectations appropriate for the declared target API status. This checklist:

- Defines the required release assets and their expected locations

> - Specifies which assets are mandatory (M) or optional (O) per API status level

Author: Applied in https://github.com/camaraproject/tooling/pull/90/commits/f81dcdca781edd417df846b73d47c6b30aed00a0

- Clarifies the division of responsibility: **codeowners prepare** the assets, **automation validates** what it can, and **release management reviewers** verify the rest

> This document replaces the deprecated per-API `API-Readiness-Checklist.md` files. Readiness is now tracked through `release-plan.yaml` configuration, preparation prerequisites in the Release Issue, and the review checklist in the Release PR.

Reviewer: I would not refer to the past approach in the new documentation.  New people do not know about the previous process.

## Terms

| Term | Source | Values |
|------|--------|--------|
| Release type | `repository.target_release_type` | `pre-release-alpha`, `pre-release-rc`, `public-release`, `maintenance-release` |
| API status | `apis[].target_api_status` | `alpha`, `rc`, `public` |
| Initial public | Derived | API status is `public` and major version is 0 (e.g., v0.5.0) |
| Stable public | Derived | API status is `public` and major version ≥ 1 (e.g., v1.0.0) |

The Release Issue and Release PR embed the readiness matrix as a convenience copy; this document is the canonical source.

## Release Assets

> | Nr | Asset | Description | Location | Automated? | How to verify |

Reviewer: ```suggestion
| Nr | Asset | Description | Location | Automated? | What to verify |
```

| 3 | Commonalities compliance | API definitions follow the Commonalities version declared in `release-plan.yaml` dependencies | In API definitions | Partially (Spectral rules cover structure; design guidelines require manual review) | Spectral covers structure; review design guideline adherence manually |
| 4 | API Documentation | API description in the YAML `info.description` field; additional documentation as needed | In YAML `info` section or `documentation/` | Partial (`info.description` presence checked) | `info.description` is present and meaningful |
| 5 | User Stories | At least one user story per API demonstrating the intended use | `documentation/API_documentation/` | No | At least one user story document per API exists |
| 6 | Test Cases (basic) | Sunny day scenarios and main error cases; at least one `.feature` file per API | `code/Test_definitions/` | File existence only | At least one `.feature` file per API with sunny day + main error cases |
| 7 | Test Cases (enhanced) | Rainy day scenarios, edge cases, and error handling coverage | `code/Test_definitions/` | No | Additional scenarios beyond basic: rainy day, edge cases, error handling |

> | 8 | API description link | Link to CAMARA Wiki API description page for external visibility | Wiki URL in API Readiness Checklist | No | Wiki URL is present and reachable |

Reviewer: For now, we could avoid hosting the URL.

## Requirements by API Status

The following matrix defines which assets are mandatory (M) or optional (O) for each target API status level:

| Nr | Asset | alpha | rc | initial public | stable public |
|----|-------|:-----:|:--:|:--------------:|:------------:|
| 1 | Release Plan | M | M | M | M |
| 2 | API Definition(s) | M | M | M | M |
| 3 | Commonalities compliance | O | M | M | M |
| 4 | API Documentation | M | M | M | M |
| 5 | User Stories | O | O | O | M |
| 6 | Test Cases (basic) | O | M | M | M |
| 7 | Test Cases (enhanced) | O | O | O | M |
| 8 | API description link | O | O | M | M |

**Why this progression:**

- **Alpha**: The API is under active development. Only the API definition and basic documentation are required. Teams are iterating on the design and gathering feedback.
- **Release Candidate (rc)**: The API is feature-complete and ready for implementation testing. Commonalities compliance and basic test cases become mandatory to ensure interoperability.

> - **Initial Public**: The API is ready for first implementations by external parties. All API versions in the release are initial (major version 0.x). An API description link is added for external visibility.

Author: Applied in https://github.com/camaraproject/tooling/pull/90/commits/f81dcdca781edd417df846b73d47c6b30aed00a0. Removed the "link" in https://github.com/camaraproject/tooling/pull/90/commits/ba596ccb73651bb298d58a2ca1a5130658cd413b

- **Release configuration matches intent**: Check that API names, target versions, target statuses, and release type in `release-plan.yaml` are correct
- **Dependency versions are current**: Commonalities and ICM dependency versions in `release-plan.yaml` should reference the latest recommended releases
- **CI checks are green**: Spectral linting and PR validation should pass on `main`
- **All intended PRs are merged**: Implementation work should be complete on `main`
- **SemVer is correct**: Breaking changes are only allowed in initial versions (v0.x) or new major versions
- **Readiness assets are provided**: All mandatory assets for the declared target release type are in place (see matrix above)

These items appear as a preparation checklist in the Release Issue while it is in PLANNED state. They are reminders, not automated gates — the codeowner is responsible for verifying them before proceeding.

> **Note**: During development on `main`, API version fields in the YAML definitions must stay as `wip`. The release automation transformer replaces them with the correct version numbers during snapshot creation.

> ## Review Process

Author: Applied in https://github.com/camaraproject/tooling/pull/90/commits/f81dcdca781edd417df846b73d47c6b30aed00a0

During the Release PR review (SNAPSHOT ACTIVE state), two types of reviewers verify different aspects:

### Codeowner Review

Codeowners verify content accuracy:

- CHANGELOG entries are organized correctly (automation collects entries from merged PRs; codeowners must move them into the correct categories and complete them)
- API version numbers match the intent declared in `release-plan.yaml`
- API definitions are correct and complete for the target status
- Test cases cover the intended API behavior
- Documentation is adequate for the target audience

### Release Management Review

Release management reviewers verify process compliance:

- CHANGELOG follows the release documentation rules
- `release-metadata.yaml` content is correct
- API version transformations were applied correctly (server URLs, references, version fields)
- Version conventions are followed (SemVer, URL format matching API maturity)

> - All mandatory readiness assets for the declared release type are present

Reviewer: ```suggestion
- All mandatory release assets for the declared API(s) status(es) are present as per the checklist
```
